File: ios/Sources/RoadSenseNSBootstrap/Network/Endpoints.swift
import Foundation

public struct Endpoints: Equatable, Sendable {
    public let config: AppConfig

    public init(config: AppConfig) {
        self.config = config
    }

    public var uploadReadingsURL: URL {
        config.functionsBaseURL.appendingPathComponent("upload-readings", isDirectory: false)
    }

    public var potholeActionsURL: URL {
        config.functionsBaseURL.appendingPathComponent("pothole-actions", isDirectory: false)
    }

    public var potholePhotosURL: URL {
        config.functionsBaseURL.appendingPathComponent("pothole-photos", isDirectory: false)
    }

    public func segmentDetailURL(id: UUID) -> URL {
        config.functionsBaseURL
            .appendingPathComponent("segments", isDirectory: true)
            .appendingPathComponent(id.uuidString.lowercased(), isDirectory: false)
    }

    public var tileTemplateURLString: String {
        let base = config.functionsBaseURL.absoluteString
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let encodedAnonKey = config.supabaseAnonKey.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)
            ?? config.supabaseAnonKey
        return base + "/tiles/{z}/{x}/{y}.mvt?apikey=" + encodedAnonKey
    }

    public func tileURL(z: Int, x: Int, y: Int, version: Int? = nil) -> URL {
        var url = config.functionsBaseURL
            .appendingPathComponent("tiles", isDirectory: true)
            .appendingPathComponent(String(z), isDirectory: true)
            .appendingPathComponent(String(x), isDirectory: true)
            .appendingPathComponent("\(y).mvt", isDirectory: false)

        if let version {
            var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
            components?.queryItems = [URLQueryItem(name: "v", value: String(version))]
            if let resolved = components?.url {
                url = resolved
            }
        }

        return url
    }
}

File: android/core-api/src/main/kotlin/ca/roadsense/ns/api/Endpoints.kt
package ca.roadsense.ns.api

import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.UUID

class Endpoints(val config: AppConfig) {
    val uploadReadingsURL: String get() = "${config.functionsBaseURL}/upload-readings"
    val potholeActionsURL: String get() = "${config.functionsBaseURL}/pothole-actions"
    val potholePhotosURL: String get() = "${config.functionsBaseURL}/pothole-photos"
    val feedbackURL: String get() = "${config.functionsBaseURL}/feedback"

    fun segmentDetailURL(id: UUID): String =
        "${config.functionsBaseURL}/segments/${id.toString().lowercase()}"

    val tileTemplateURLString: String
        get() {
            val encodedAnonKey = URLEncoder.encode(config.supabaseAnonKey, StandardCharsets.UTF_8)
            val base = config.functionsBaseURL.trimEnd('/')
            return "$base/tiles/{z}/{x}/{y}.mvt?apikey=$encodedAnonKey"
        }

    fun tileURL(z: Int, x: Int, y: Int, version: Int? = null): String {
        val base = "${config.functionsBaseURL}/tiles/$z/$x/$y.mvt"
        return if (version == null) base else "$base?v=$version"
    }
}
